File: dotnet/examples/Pactman.NonprofitCheckPlus.Examples/Cases/Ex15MalformedEin.cs
using System;
using System.Linq;
using System.Threading.Tasks;
using Pactman.NonprofitCheckPlus.Exceptions;

namespace Pactman.NonprofitCheckPlus.Examples.Cases;

/// <summary>
/// EX-15 — Malformed EIN rejected locally.
/// </summary>
/// <remarks>
/// A malformed EIN never reaches the network, never spends quota, and reports which item
/// failed and why.
/// </remarks>
public sealed class Ex15MalformedEin : IExample
{
    /// <inheritdoc />
    public string Id => "ex-15";

    /// <inheritdoc />
    public string Title => "A malformed EIN rejected before the request, spending nothing.";

    /// <inheritdoc />
    public async Task RunAsync(string[] args)
    {
        using var context = ExampleContext.Live();

        Output.Heading("A single malformed EIN");

        try
        {
            await context.Client.Nonprofits.CheckAsync("not-an-ein");

            throw new ExampleFailedException("A malformed EIN was accepted. It should have been rejected.");
        }
        catch (PactmanValidationException error)
        {
            Output.Field("type", error.GetType().Name);
            Output.Field("category", error.Category.ToWireValue());

            // Local, not api: nothing was sent, so nothing was billed and there is no
            // request id to chase.
            Output.Field("origin", error.Origin.ToWireValue());
            Output.Field("message", error.Message);

            var issue = error.Issues.Single();
            Output.Field("issue.index", issue.Index);
            Output.Field("issue.value", issue.Value);
        }

        Output.Heading("A batch where some are malformed");

        try
        {
            await context.Client.Nonprofits.CheckBulkAsync(
                new[] { "41-1787097", "nope", "996589560", "12345" });

            throw new ExampleFailedException("A malformed batch was accepted. It should have been rejected.");
        }
        catch (PactmanValidationException error)
        {
            // Every failure at once. A caller fixes the whole batch in one pass rather
            // than one item per round trip.
            Output.Field("message", error.Message);

            foreach (var issue in error.Issues)
            {
                Output.Field($"index {issue.Index}", $"{issue.Value} — {issue.Message}");
            }
        }

        Output.Heading("Asking without throwing");

        foreach (var candidate in new[] { "41-1787097", "nope" })
        {
            Output.Field(candidate, Ein.IsValid(candidate) ? "valid" : "invalid");
        }

        Output.Note(
            "Validation is local and free. It is also formatting-only: a well-formed EIN\n"
            + "may still have no record, which is EX-16 and a different thing entirely.");
    }
}

File: java/examples/src/main/java/org/pactman/nonprofitcheckplus/examples/cases/Ex15MalformedEin.java
package org.pactman.nonprofitcheckplus.examples.cases;

import java.util.Arrays;
import org.pactman.nonprofitcheckplus.examples.support.Example;
import org.pactman.nonprofitcheckplus.examples.support.ExampleContext;
import org.pactman.nonprofitcheckplus.examples.support.ExampleFailedException;
import org.pactman.nonprofitcheckplus.examples.support.Output;
import org.pactman.nonprofitcheckplus.exceptions.ErrorOrigin;
import org.pactman.nonprofitcheckplus.exceptions.PactmanValidationException;
import org.pactman.nonprofitcheckplus.exceptions.ValidationIssue;

/**
 * EX-15 — A malformed EIN never reaches the API.
 *
 * <p>Local validation is not a convenience. Every request is billable, and a
 * request that cannot succeed should not be one of them.
 */
public final class Ex15MalformedEin implements Example {

    @Override
    public String id() {
        return "ex-15";
    }

    @Override
    public String title() {
        return "Malformed input fails locally, before a billable request";
    }

    @Override
    public void run(String[] args) {
        try (ExampleContext context = ExampleContext.withFixtures()) {
            Output.heading("A single malformed EIN");

            try {
                context.client().nonprofits().check("4117870");
                throw new ExampleFailedException("The malformed EIN should have been rejected.");
            } catch (PactmanValidationException failure) {
                Output.field("exception", failure.getClass().getSimpleName());
                Output.field("category", failure.category());
                Output.field("origin", failure.origin());
                Output.field("message", failure.getMessage());
            }

            Output.heading("A batch with two bad rows");

            try {
                context.client()
                        .nonprofits()
                        .checkBulk(Arrays.asList("411787097", "nope", "996589560", "1234"));
                throw new ExampleFailedException("The batch should have been rejected.");
            } catch (PactmanValidationException failure) {
                Output.field("message", failure.getMessage());
                System.out.println();

                for (ValidationIssue issue : failure.issues()) {
                    Output.bullet("index " + issue.index() + " (" + issue.value() + "): "
                            + issue.message());

                    if (issue.index() == null) {
                        throw new ExampleFailedException(
                                "A batch issue should name the row it came from.");
                    }
                }

                if (failure.origin() != ErrorOrigin.LOCAL) {
                    throw new ExampleFailedException("Local validation should have origin LOCAL.");
                }
            }
        }

        Output.note("origin is what separates these from an API-side rejection. A\n"
                + "PactmanValidationException with origin LOCAL means nothing was sent and\n"
                + "nothing was billed; a PactmanBadRequestException with origin API means the\n"
                + "API looked at the request and refused it. Retrying the first is pointless\n"
                + "until the input changes, which is why neither is ever retried.");
    }
}
